Lib/typing.py

```diff
@@ -521,13 +521,26 @@ def _evaluate(self, globalns, localns):
             self.__forward_evaluated__ = True
         return self.__forward_value__
 
+    def _check_recursion(self, value, forward_args):
+        if isinstance(value, _GenericAlias):
+            for val in value.__args__:
+                if isinstance(val, _GenericAlias):
+                    self._check_recursion(val, forward_args)
+                elif self is val and self.__forward_evaluated__:
+                    forward_args.append(...)
```

Comment:
Is it possible to create an actually recursive object? Or is it too dangerous?

Reply:
What does an actually recursive object mean? I added some test code in new commit. Can you have a look and see what other tests need to be added. Thanks.

Reply:
```
a = List['int']
for i in range(sys.getrecursionlimit() + 100):
    a = List[a]
```
It raises a `RecursionError` both in this PR and master branch. Is an actually recursive object mean this?

Reply:
No. Consider the initial example:
```python
ValueList = List['Value']
Value = Union[str, ValueList]
```
In this example I mean trying to construct an object `Value` such that `Value.__args__[1].__args__[0] is Value`. I understand this may be tricky, but I don't think it is impossible.

Also it indeed looks like this may be "dangerous" in the sense that we need to make `__repr__()`, `__eq__()`, and `__hash__()` aware of the possible recursion.

Reply:
On the other hand:
```
>>> x = [1]
>>> x.append(x)
>>> x
[1, [...]]
>>> y = [1]
>>> y.append(y)
>>> x == y
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RecursionError: maximum recursion depth exceeded in comparison
```
So maybe it would be enough to just update the `__repr__()`.